src/ivc/cyclefold/incrementally_verifiable_computation/mod.rs

```diff
@@ -52,10 +74,522 @@ where
+                z_i: z_0,
+                z_0,
+            }
+            .build(),
+            _p: PhantomData,
+        };
+
+        let primary_initial_instances = primary_sfc.initial_instances();
+
+        #[cfg(test)]
+        {
+            let _mock = info_span!("mock_debug").entered();
+            crate::halo2_proofs::dev::MockProver::run(
+                pp.primary_k_table_size,
+                &primary_sfc,
+                primary_initial_instances.clone(),
+            )
+            .unwrap()
+            .verify()
+            .unwrap();
+        }
+
+        let primary_cr = CircuitRunner::new(
+            pp.primary_k_table_size,
+            primary_sfc,
+            primary_initial_instances.clone(),
+        );
+        let primary_witness = primary_cr
+            .try_collect_witness()
+            .map_err(|err| Error::WhileCollectPrimaryWitness { step: 0, err })?;
+        pp.primary_S = primary_cr
+            .try_collect_plonk_structure()
+            .map_err(|err| Error::WhileCollectPrimaryS { err })?;
+
+        let primary_post_initial_trace = ProtoGalaxy::<CMain, 1>::generate_plonk_trace(
+            &pp.primary_ck,
+            &primary_initial_instances,
+            &primary_witness,
+            &pp.protogalaxy_prover_params(),
+            &mut ro(),
+        )?;
+
+        Ok(Self {
+            step: NonZeroUsize::new(1).expect("safe: 1 != 0"),
+            // Because zero step using input values for output without any folding (only formal
+            // on-circuit) - we just take initial acc-s & z_0
+            primary_z_current: z_0,
+            primary_z_0: z_0,
+            primary_trace: primary_post_initial_trace,
+            primary_acc: primary_initial_acc,
+            support_acc: support_initial_acc,
+            _p: PhantomData,
+        })
+    }
+
+    pub fn next(
+        self,
+        pp: &PublicParams<ARITY, CMain, CSup, SC>,
+        sc: &SC,
+    ) -> Result<Self, Error<CMain>> {
+        let _span = info_span!("ivc_next", step = self.step.get()).entered();
+
+        let Self {
+            step,
+            primary_acc,
+            primary_trace,
+            primary_z_current,
+            primary_z_0,
+            support_acc,
+            _p,
+        } = self;
+
+        let mut random_oracle = ro();
+        let (primary_next_acc, primary_proof) = ProtoGalaxy::prove(
+            &pp.primary_ck,
+            &pp.protogalaxy_prover_params(),
+            &mut random_oracle,
+            primary_acc.clone(),
+            &[primary_trace.clone()],
+        )
+        .unwrap();
+
+        // Inside `prove` we use several challenges generated from `random oracle`. Since we
+        // delegate the `W_commitment` calculations to the `upport_circuit` we need `gamma` to
+        // repeat them. We take advantage of the fact that `gamma` is generated last and thus
+        // retrieve it
+        let gamma = random_oracle.squeeze::<CMain::ScalarExt>(MAX_BITS);
```

Comment:
is gamma already  squeezed in `fn ProtoGalaxy::prove`?
Here it squeeze a new challenge and hence will be different from the gamma generated in `ProtoGalaxy::prove`. 
Why are they the same?

Reply:
Random Oracle does not clear itself after generating a challenge, so if you pass it in and then ask it to generate again, it will return the same value

I use this so that I don't have to create an additional API to extract this gamma

Reply:
do you mean if we absorb some input data `ro.absorb(input_data)` and call the squeeze multiple times:
```
let r1 = ro.squeeze();
let r2 = ro.squeeze();
```
r1 and r2 will be the same?